tests/sensitivity_analysis_test.py

```diff
@@ -0,0 +1,303 @@
+    net = a.train(training_batch_size=16, stop_after_epochs=1, max_num_epochs=2)
+    assert isinstance(net, nn.Module)
+
+
+def test_add_property_invalid_model_raises(toy_theta_property, posterior_stub) -> None:
+    theta, y = toy_theta_property
+    a = ActiveSubspace(posterior_stub)
+    with pytest.raises(NameError):
+        a.add_property(theta=theta, emergent_property=y, model="unknown")
+
+
+def test_train_reuses_existing_net(toy_theta_property, posterior_stub) -> None:
+    theta, y = toy_theta_property
+    a = ActiveSubspace(posterior_stub)
+    a.add_property(theta=theta, emergent_property=y, model="mlp", hidden_features=8)
+    _ = a.train(training_batch_size=16, stop_after_epochs=1, max_num_epochs=2)
+    first_id = id(a._regression_net)
+    _ = a.train(training_batch_size=16, stop_after_epochs=1, max_num_epochs=2)
+    assert id(a._regression_net) == first_id  # net is reused, not rebuilt
+
+
+# ------------------------
+# ActiveSubspace.find_directions tests
+# ------------------------
+
+
+@pytest.mark.parametrize("norm_gradients", [True, False])
+def test_find_directions_with_regression_net(
+    norm_gradients, toy_theta_property, posterior_stub
+) -> None:
+    """Tests that find_directions runs and returns correctly shaped outputs."""
+    theta, y = toy_theta_property
+    a = ActiveSubspace(posterior_stub)
+    a.add_property(theta=theta, emergent_property=y, model="mlp", hidden_features=8)
+    a.train(training_batch_size=16, stop_after_epochs=1, max_num_epochs=2)
+
+    evals, evecs = a.find_directions(
+        posterior_log_prob_as_property=False,
+        norm_gradients_to_prior=norm_gradients,
+        num_monte_carlo_samples=128,
+    )
+    d = theta.shape[1]
+    assert evals.shape == (d,)
+    assert evecs.shape == (d, d)
+    # Ascending order
+    assert torch.all(evals[1:] >= evals[:-1])
+    # Columns are unit vectors
+    assert torch.allclose(torch.linalg.norm(evecs, dim=0), torch.ones(d), atol=1e-5)
+
+
+def test_find_directions_with_posterior_log_prob_warns(
+    toy_theta_property, posterior_stub
+) -> None:
+    theta, y = toy_theta_property
+    a = ActiveSubspace(posterior_stub)
+    a.add_property(theta=theta, emergent_property=y, model="mlp", hidden_features=8)
+    a.train(training_batch_size=16, stop_after_epochs=1, max_num_epochs=2)
+
+    with pytest.warns(UserWarning):
+        evals, evecs = a.find_directions(
+            posterior_log_prob_as_property=True,
+            norm_gradients_to_prior=True,
+            num_monte_carlo_samples=64,
+        )
+    assert evals.numel() == evecs.shape[0]
+
+
+def test_find_directions_raises_without_property(posterior_stub) -> None:
+    a = ActiveSubspace(posterior_stub)
+    with pytest.raises(ValueError):
+        _ = a.find_directions(
+            posterior_log_prob_as_property=False, num_monte_carlo_samples=8
+        )
+
+
+# ------------------------
+# ActiveSubspace.project tests
+# ------------------------
```

Comment:
what is .project?

Reply:
it's the `ActiveSubspace.project(...)` method